packages/stateManager/index.ts

```diff
@@ -1,104 +1,636 @@
+      const ui = event.updateItem;
+      const id = bytesToHex(ui.itemId);
+      const item = await this.store.get(id);
+      if (ui.metadata) {
+        item.metadata = bufferToJSON(ui.metadata);
+      }
+      if (ui.price) {
+        item.price = ui.price;
+      }
+      await this.store.put(id, item);
+      this.emit("updateItem", item);
+      return;
+    } else if (event.changeStock) {
+      const cs = event.changeStock;
+      if (cs.itemIds) {
+        await Promise.all(
+          cs.itemIds.map(async (id: Uint8Array, i: number) => {
+            const itemId = bytesToHex(id);
+            const item = await this.store.get(itemId);
+            const diff = cs.diffs ? cs.diffs[i] : 0;
+            item.quantity = item.quantity + diff;
+            await this.store.put(itemId, item);
+          }),
+        );
+        this.emit("changeStock", { eventId: cs.eventId });
+        return;
+      }
+    } else if (event.updateTag) {
+      // Add or remove tagId to item
+      const ut = event.updateTag;
+      const tagId = bytesToHex(ut.tagId);
 
-class ItemManager extends PublicObjectManager<Item> {
-  constructor(db: Store<Item>, client: RelayClient) {
-    super(db, client);
-  }
-  // Process the events; if the event modifies one the shops items this updates the store and emits an event
-  // returns a bool to indicate whether the event was processed or not
-  async _processEvent(event: schema.ShopEvents): Promise<boolean> {
-    if (typeof event === schema.ShopEvents.UpdateItem) {
-      // load the item from the store
-      const item = await this.db.get(event.eventId);
-      Object.assign(item, event);
-      await this.db.put(event.eventId, item);
-      this.emit("update", item);
-      return true;
-    } else {
-      return false;
+      if (ut.addItemId) {
+        const itemId = bytesToHex(ut.addItemId);
+        const item = await this.store.get(itemId);
+        item.tags.push(tagId);
+        await this.store.put(itemId, item);
+        this.emit("addItemId", item);
+        return;
+      }
+      if (ut.removeItemId) {
+        const itemId = bytesToHex(ut.removeItemId);
+        const item = await this.store.get(itemId);
+        // remove `tagId` from item.tags array
+        item.tags = [...item.tags.filter((id: `0x${string}`) => id !== tagId)];
+        await this.store.put(itemId, item);
+        this.emit("removeItemId", item);
+        return;
+      }
     }
   }
-  create(item: Item) {
-    return Promise.all([
-      this.db.put(item.id, item),
-      this.client.createItem(item),
-    ]);
+
+  async create(item: Partial<Item>): Promise<Item> {
+    const eventId = await this.client.createItem({
+      price: item.price,
+      metadata: convertToBuffer(item.metadata),
+    });
+
+    return new Promise((resolve, reject) => {
+      const remove = () => this.removeListener("createItem", onCreate);
+      function onCreate(create: Item) {
+        if (create.id === bytesToHex(eventId)) {
```

Comment:
it isn't immediately obvious why we compare against `create.id` here, since that field doesn't exist on the network/protobuf level.

in general i'd advice to keep the name for the ID-field every wehre the same, as much as possible.

Then we should also be possible to turn this promise-waits-for-event construction into a  function that is re-usable across stores, by only passing in the event type and id in.